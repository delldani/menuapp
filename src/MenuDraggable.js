import React, { Component } from 'react';

export function sum(a, b) {
  return a + b;
}

export default class MenuDraggable extends Component {
  constructor(props) {
    super(props);
    this.ondragend = this.ondragend.bind(this);
    this.valami = this.valami.bind(this);

    this.oneDimArray = JSON.parse(JSON.stringify(this.props.menuItems));
    this.state = {
      menuItemsArray: this.sortElements(this.oneDimArray),
      active: null,
      activeItemIndex: null,
      number: 0,
    };
  }

  sortElements(oneDimArray) {
    let array = JSON.parse(JSON.stringify(oneDimArray));

    function sort(item) {
      array.map((elem) => {
        if (item.name == elem.parent) {
          if (!item.hasOwnProperty('children')) {
            item.children = [];
          }
          item.children.push(elem);
          sort(elem);
        }
      });
    }

    const result = array.filter((item) => item.parent == '');
    result.map((item) => {
      sort(item);
    });

    return result;
  }

  allowDrop(ev) {
    ev.preventDefault();
  }

  dragStart = (index) => (ev) => {
    ev.stopPropagation();
    ev.dataTransfer.setData('text/plain', ev.target);
    this.setState({ active: 'moving', activeItemIndex: index, number: 55 });
  };

  valami(a) {
    return a;
  }
  drop = (targetId) => (ev) => {
    ev.preventDefault();
    ev.stopPropagation();

    if (this.state.active) {
      const itemIndex = this.oneDimArray.findIndex(
        (item) => item.name == this.state.activeItemIndex,
      );
      const targetIndex = this.oneDimArray.findIndex((item) => item.name == targetId);

      if (!(this.oneDimArray[targetIndex].name == this.oneDimArray[itemIndex].parent)) {
        const temp = this.oneDimArray[targetIndex].name;
        const color = this.oneDimArray[targetIndex].color;
        this.oneDimArray[targetIndex].name = this.oneDimArray[itemIndex].name;
        this.oneDimArray[itemIndex].name = temp;
        this.oneDimArray[targetIndex].color = this.oneDimArray[itemIndex].color;
        this.oneDimArray[itemIndex].color = color;

        const t = this.sortElements(this.oneDimArray);

        this.setState({ menuItemsArray: t }, () => {});
        // this.setState({ activeItemIndex: targetIndex });
      }
    }
  };

  ondragenter = (targetId) => (ev) => {
    ev.preventDefault();
  };

  ondragend(ev) {
    this.setState({ active: null, activeItemIndex: null });
  }

  getItems() {
    let newArray = [];

    const dragStart = this.dragStart;
    const onDrop = this.drop;
    const onDragOver = this.onDragOver;
    const onDragEnd = this.onDragEnd;

    function makeTree(item) {
      if (!item.hasOwnProperty('children')) {
        return;
      } else {
        return (
          <ul style={{ listStyleType: 'none' }}>
            {item.children.map((item, index) => {
              return (
                <li
                  id={item.name}
                  key={index}
                  draggable="true"
                  style={{
                    backgroundColor: item.color,
                    width: 90,
                    height: 30,
                    textAlign: 'center',
                  }}
                  onDragStart={dragStart(item.name)}
                  onDrop={onDrop(item.name)}
                  onDragOver={onDragOver}
                  onDragEnd={onDragEnd}
                >
                  {item.name}
                  <br></br>
                  {makeTree(item)}
                </li>
              );
            })}
          </ul>
        );
      }
    }

    let arrayFromState = JSON.parse(JSON.stringify(this.state.menuItemsArray));
    newArray = arrayFromState.map((item, index) => {
      return (
        <li
          id={item.name}
          key={index}
          style={{ backgroundColor: item.color, width: 90, height: 30, textAlign: 'center' }}
          draggable="true"
          onDragStart={this.dragStart(item.name)}
          onDrop={this.drop(item.name)}
          onDragOver={this.allowDrop}
          onDragEnd={this.ondragend}
          // onDragEnter={this.ondragenter(item.name)}
          onClick={() => this.valami(2)}
        >
          {item.name}
          <br></br>
          {makeTree(item)}
        </li>
      );
    });
    return newArray;
  }

  onClick = (e) => {
    const newNumber = ++this.state.number;
    this.setState({ number: newNumber });
  };

  render() {
    const items = this.getItems();

    return (
      <div>
        <button id="b1" onClick={this.onClick}>
          {this.state.number}
        </button>
        sasa
        <ul style={{ listStyleType: 'none' }}>{items}</ul>
      </div>
    );
  }
}
