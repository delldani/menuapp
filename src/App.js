import React, { Component } from 'react';

import MenuDraggable from './MenuDraggable';

const menu1 = [
  { name: 'menu 1', child: '', color: 'blue' },
  { name: 'menu 2', child: '', color: 'red' },
  { name: 'menu 3', child: 'menu 5', color: 'green' },
  { name: 'menu 4', child: '', color: 'yellow' },
  { name: 'menu 5', child: '', color: 'pink' },
];
const menu2 = [
  { name: 'gizu1', child: '', color: 'blue' },
  { name: 'bali2', child: '', color: 'red' },
  { name: 'jani3', child: '', color: 'green' },
  { name: 'dani4', child: '', color: 'yellow' },
];

function App() {
  return (
    <div>
      <MenuDraggable menuItems={menu1} />
      <MenuDraggable menuItems={menu2} />
    </div>
  );
}

export default App;
