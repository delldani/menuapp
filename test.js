import { sum } from './src/MenuDraggable';
import MenuDraggable from './src/MenuDraggable';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React, { Component } from 'react';
import { shallow, mount } from 'enzyme';
import { wrap } from 'module';

Enzyme.configure({ adapter: new Adapter() });
const menu1 = [
  { name: 'menu 1', parent: '', color: 'blue' },
  { name: 'menu 2', parent: '', color: 'red' },
  { name: 'menu 4', parent: '', color: 'yellow' },
  { name: 'menu 5', parent: '', color: 'pink' },
  { name: 'menu 6', parent: 'menu 4', color: 'pink' },
];
const expectedMenu = [
  { name: 'menu 1', parent: '', color: 'blue' },
  { name: 'menu 2', parent: '', color: 'red' },
  {
    name: 'menu 4',
    parent: '',
    color: 'yellow',
    children: [
      {
        color: 'pink',
        name: 'menu 6',
        parent: 'menu 4',
      },
    ],
  },
  { name: 'menu 5', parent: '', color: 'pink' },
];

test('adds 1 + 2 to equal 3', () => {
  expect(sum(1, 2)).toBe(3);
});

test('enzyme test shallow rendering', () => {
  const wrapper = shallow(<MenuDraggable menuItems={menu1} />);
  expect(wrapper.contains('sasa')).toBe(true);
});

test('enzyme test DOM rendering', () => {
  const wrapper = mount(<MenuDraggable menuItems={menu1} />);
  expect(wrapper.instance().sortElements(menu1)).toEqual(expectedMenu);
  expect(wrapper.state().active).toBe(null);
  wrapper.instance().dragStart('menu 1');
  wrapper.find([(className = 'menu 2')]).simulate('click');
  expect(wrapper.state().active).toBe('moving');
});
