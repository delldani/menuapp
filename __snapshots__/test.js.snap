// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react testing library 1`] = `
<div>
  <div>
    <button
      id="b1"
    >
      0
    </button>
    sasa
    <ul
      style="list-style-type: none;"
    >
      <li
        draggable="true"
        id="menu 1"
        style="background-color: blue; width: 90px; height: 30px; text-align: center;"
      >
        menu 1
        <br />
      </li>
      <li
        draggable="true"
        id="menu 2"
        style="background-color: red; width: 90px; height: 30px; text-align: center;"
      >
        menu 2
        <br />
      </li>
      <li
        draggable="true"
        id="menu 4"
        style="background-color: yellow; width: 90px; height: 30px; text-align: center;"
      >
        menu 4
        <br />
        <ul
          style="list-style-type: none;"
        >
          <li
            draggable="true"
            id="menu 6"
            style="background-color: pink; width: 90px; height: 30px; text-align: center;"
          >
            menu 6
            <br />
          </li>
        </ul>
      </li>
      <li
        draggable="true"
        id="menu 5"
        style="background-color: pink; width: 90px; height: 30px; text-align: center;"
      >
        menu 5
        <br />
      </li>
    </ul>
  </div>
</div>
`;

exports[`react testing library 2`] = `
<div>
  <div>
    <button
      id="b1"
    >
      55
    </button>
    sasa
    <ul
      style="list-style-type: none;"
    >
      <li
        draggable="true"
        id="menu 2"
        style="background-color: red; width: 90px; height: 30px; text-align: center;"
      >
        menu 2
        <br />
      </li>
      <li
        draggable="true"
        id="menu 1"
        style="background-color: blue; width: 90px; height: 30px; text-align: center;"
      >
        menu 1
        <br />
      </li>
      <li
        draggable="true"
        id="menu 4"
        style="background-color: yellow; width: 90px; height: 30px; text-align: center;"
      >
        menu 4
        <br />
        <ul
          style="list-style-type: none;"
        >
          <li
            draggable="true"
            id="menu 6"
            style="background-color: pink; width: 90px; height: 30px; text-align: center;"
          >
            menu 6
            <br />
          </li>
        </ul>
      </li>
      <li
        draggable="true"
        id="menu 5"
        style="background-color: pink; width: 90px; height: 30px; text-align: center;"
      >
        menu 5
        <br />
      </li>
    </ul>
  </div>
</div>
`;
